File: frontend/src/API/productPriceService.js
import api from './client';

const productPriceService = {
  /**
   * Get all product prices - requires moderator role
   * @returns {Promise<Array>} List of product price objects
   */
  getAllProductPrices: async () => {
    try {
      const response = await api.get('/api/analytics/moderator/prices');
      return response.data;
    } catch (error) {
      console.error('Error fetching all product prices:', error);
      throw error;
    }
  },

  /**
   * Get active product prices - public endpoint
   * @returns {Promise<Array>} List of active product price objects
   */
  getActiveProductPrices: async () => {
    try {
      const response = await api.get('/api/analytics/prices');
      return response.data;
    } catch (error) {
      console.error('Error fetching active product prices:', error);
      throw error;
    }
  },

  /**
   * Get product prices by category - public endpoint
   * @param {string} category - The category to filter by
   * @returns {Promise<Array>} List of product price objects in the specified category
   */
  getProductPricesByCategory: async (category) => {
    try {
      const response = await api.get(`/api/analytics/prices/category/${category}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching product prices for category ${category}:`, error);
      throw error;
    }
  },

  /**
   * Get product price by ID - requires moderator role
   * @param {number} id - The ID of the product price to fetch
   * @returns {Promise<Object>} The product price object
   */
  getProductPriceById: async (id) => {
    try {
      const response = await api.get(`/api/analytics/moderator/prices/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching product price with ID ${id}:`, error);
      throw error;
    }
  },

  /**
   * Create a new product price - requires moderator role
   * @param {Object} priceData - The product price data
   * @returns {Promise<Object>} The created product price object
   */
  createProductPrice: async (priceData) => {
    try {
      const response = await api.post('/api/analytics/moderator/prices', priceData);
      return response.data;
    } catch (error) {
      console.error('Error creating product price:', error);
      throw error;
    }
  },

  /**
   * Update an existing product price - requires moderator role
   * @param {number} id - The ID of the product price to update
   * @param {Object} priceData - The updated product price data
   * @returns {Promise<Object>} The updated product price object
   */
  updateProductPrice: async (id, priceData) => {
    try {
      const response = await api.put(`/api/analytics/moderator/prices/${id}`, priceData);
      return response.data;
    } catch (error) {
      console.error(`Error updating product price with ID ${id}:`, error);
      throw error;
    }
  },

  /**
   * Delete a product price - requires moderator role
   * @param {number} id - The ID of the product price to delete
   * @returns {Promise<void>}
   */
  deleteProductPrice: async (id) => {
    try {
      await api.delete(`/api/analytics/moderator/prices/${id}`);
    } catch (error) {
      console.error(`Error deleting product price with ID ${id}:`, error);
      throw error;
    }
  },

  /**
   * Deactivate a product price - requires moderator role
   * @param {number} id - The ID of the product price to deactivate
   * @returns {Promise<Object>} The updated product price object
   */
  deactivateProductPrice: async (id) => {
    try {
      const response = await api.put(`/api/analytics/moderator/prices/${id}/deactivate`);
      return response.data;
    } catch (error) {
      console.error(`Error deactivating product price with ID ${id}:`, error);
      throw error;
    }
  },

  /**
   * Activate a product price - requires moderator role
   * @param {number} id - The ID of the product price to activate
   * @returns {Promise<Object>} The updated product price object
   */
  activateProductPrice: async (id) => {
    try {
      const response = await api.put(`/api/analytics/moderator/prices/${id}/activate`);
      return response.data;
    } catch (error) {
      console.error(`Error activating product price with ID ${id}:`, error);
      throw error;
    }
  }
};

export default productPriceService;

File: backend/analytic-service/src/main/java/com/springcloud/service/ProductPriceService.java
package com.springcloud.service;

import com.springcloud.dto.ProductPriceDTO;
import com.springcloud.dto.ProductPriceRequest;
import com.springcloud.model.ProductPrice;
import com.springcloud.repository.ProductPriceRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class ProductPriceService {

    @Autowired
    private ProductPriceRepository productPriceRepository;

    /**
     * Get all product prices
     */
    public List<ProductPriceDTO> getAllProductPrices() {
        return productPriceRepository.findAll().stream()
                .map(this::mapToDTO)
                .collect(Collectors.toList());
    }

    /**
     * Get active product prices
     */
    public List<ProductPriceDTO> getActiveProductPrices() {
        return productPriceRepository.findByActive(true).stream()
                .map(this::mapToDTO)
                .collect(Collectors.toList());
    }

    /**
     * Get product prices by category
     */
    public List<ProductPriceDTO> getProductPricesByCategory(String category) {
        return productPriceRepository.findByCategory(category).stream()
                .map(this::mapToDTO)
                .collect(Collectors.toList());
    }

    /**
     * Get product price by ID
     */
    public Optional<ProductPriceDTO> getProductPriceById(Long id) {
        return productPriceRepository.findById(id)
                .map(this::mapToDTO);
    }

    /**
     * Create new product price
     */
    @Transactional
    public ProductPriceDTO createProductPrice(ProductPriceRequest request) {
        ProductPrice productPrice = new ProductPrice();
        updateEntityFromRequest(productPrice, request);
        productPrice = productPriceRepository.save(productPrice);
        return mapToDTO(productPrice);
    }

    /**
     * Update product price
     */
    @Transactional
    public Optional<ProductPriceDTO> updateProductPrice(Long id, ProductPriceRequest request) {
        Optional<ProductPrice> optionalProductPrice = productPriceRepository.findById(id);
        if (optionalProductPrice.isPresent()) {
            ProductPrice productPrice = optionalProductPrice.get();
            updateEntityFromRequest(productPrice, request);
            productPrice = productPriceRepository.save(productPrice);
            return Optional.of(mapToDTO(productPrice));
        }
        return Optional.empty();
    }

    /**
     * Delete product price
     */
    @Transactional
    public boolean deleteProductPrice(Long id) {
        if (productPriceRepository.existsById(id)) {
            productPriceRepository.deleteById(id);
            return true;
        }
        return false;
    }

    /**
     * Deactivate product price
     */
    @Transactional
    public Optional<ProductPriceDTO> deactivateProductPrice(Long id) {
        Optional<ProductPrice> optionalProductPrice = productPriceRepository.findById(id);
        if (optionalProductPrice.isPresent()) {
            ProductPrice productPrice = optionalProductPrice.get();
            productPrice.setActive(false);
            productPrice = productPriceRepository.save(productPrice);
            return Optional.of(mapToDTO(productPrice));
        }
        return Optional.empty();
    }

    /**
     * Activate product price
     */
    @Transactional
    public Optional<ProductPriceDTO> activateProductPrice(Long id) {
        Optional<ProductPrice> optionalProductPrice = productPriceRepository.findById(id);
        if (optionalProductPrice.isPresent()) {
            ProductPrice productPrice = optionalProductPrice.get();
            productPrice.setActive(true);
            productPrice = productPriceRepository.save(productPrice);
            return Optional.of(mapToDTO(productPrice));
        }
        return Optional.empty();
    }

    /**
     * Update entity from request
     */
    private void updateEntityFromRequest(ProductPrice productPrice, ProductPriceRequest request) {
        productPrice.setProductName(request.getProductName());
        productPrice.setCategory(request.getCategory());
        productPrice.setMinPrice(request.getMinPrice());
        productPrice.setMaxPrice(request.getMaxPrice());
        productPrice.setRecommendedPrice(request.getRecommendedPrice());
        productPrice.setUnit(request.getUnit());
        productPrice.setDescription(request.getDescription());
        if (request.getActive() != null) {
            productPrice.setActive(request.getActive());
        }
    }

    /**
     * Map entity to DTO
     */
    private ProductPriceDTO mapToDTO(ProductPrice productPrice) {
        ProductPriceDTO dto = new ProductPriceDTO();
        dto.setId(productPrice.getId());
        dto.setProductName(productPrice.getProductName());
        dto.setCategory(productPrice.getCategory());
        dto.setMinPrice(productPrice.getMinPrice());
        dto.setMaxPrice(productPrice.getMaxPrice());
        dto.setRecommendedPrice(productPrice.getRecommendedPrice());
        dto.setUnit(productPrice.getUnit());
        dto.setDescription(productPrice.getDescription());
        dto.setActive(productPrice.getActive());
        dto.setCreatedAt(productPrice.getCreatedAt());
        dto.setUpdatedAt(productPrice.getUpdatedAt());
        return dto;
    }
}
